server/internal/infprocessor/pipe.go

```diff
@@ -28,6 +28,9 @@ func (c pipeReadWriteCloser) Close() error {
 
 // closeWrite closes the write pipe.
 func (c pipeReadWriteCloser) closeWrite() error {
+	if c.PipeWriter == nil {
```

Comment:
How about changing `task.bodyWriter` to ` *pipeReadWriteCloser` and add a nil-check before calling `t.bodyWriter.closeWrite()`?

```
diff --git a/server/internal/infprocessor/infprocessor.go b/server/internal/infprocessor/infprocessor.go
index cff8edb..51ba0fe 100644
--- a/server/internal/infprocessor/infprocessor.go
+++ b/server/internal/infprocessor/infprocessor.go
@@ -38,7 +38,7 @@ type task struct {
 	Header http.Header
 
 	setResult  func(*result)
-	bodyWriter pipeReadWriteCloser
+	bodyWriter
 
 	EngineID string
 
@@ -392,8 +392,10 @@ func (p *P) RemoveEngine(engineID string, clusterInfo *auth.ClusterInfo) {
 		p.logger.Info("Canceled task", "reason", "engine removed", "engine", engineID, "task", t.ID)
 		delete(p.inProgressTasksByID, t.ID)
 		t.setResult(&result{err: fmt.Errorf("engine %s is removed", engineID), retry: true})
-		if err := t.bodyWriter.closeWrite(); err != nil {
-			p.logger.Error(err, "Failed to close the body writer when engine removed", "engine", t.EngineID, "task", t.ID)
+		if t.bodyWriter != nil {
+			if err := t.bodyWriter.closeWrite(); err != nil {
+				p.logger.Error(err, "Failed to close the body writer when engine removed", "engine", t.EngineID, "task", t.ID)
+			}
 		}
 	}
 
diff --git a/server/internal/infprocessor/pipe.go b/server/internal/infprocessor/pipe.go
index 86090f4..6482407 100644
--- a/server/internal/infprocessor/pipe.go
+++ b/server/internal/infprocessor/pipe.go
@@ -2,9 +2,9 @@ package infprocessor
 
 import "io"
 
-func newPipeReadWriteCloser() pipeReadWriteCloser {
+func newPipeReadWriteCloser() *pipeReadWriteCloser {
 	pr, pw := io.Pipe()
-	return pipeReadWriteCloser{
+	return &pipeReadWriteCloser{
 		PipeReader: pr,
 		PipeWriter: pw,
 	}
@@ -28,9 +28,6 @@ func (c pipeReadWriteCloser) Close() error {
 
 // closeWrite closes the write pipe.
 func (c pipeReadWriteCloser) closeWrite() error {
-	if c.PipeWriter == nil {
-		return nil
-	}
 	if err := c.PipeWriter.Close(); err != nil {
 		return err
 	}
```